File: monad/state.hs
-- state monad

import Control.Monad.State

bind' :: MonadState t1 m => State t1 t -> (t -> State t1 a) -> m a
bind' a b = state $ \s ->
            let (r1, s1) = runState a      s
                (r2, s2) = runState (b r1) s1
            in  (r2, s2)

return' :: MonadState s m => a -> m a
return' x = state $ \s -> (x, s)

get'      = state $ \s -> (s, s)

put'    x = state $ \_ -> ((), x)

modify' f = state $ \s -> ((), f s)

main = do
  print $ runState (return' 100) 10 -- (100,10)
  print $ (`runState` 0) $ return' 10 `bind'` (\x -> state $ \s -> (x, 2000)) -- (10,2000)
  print $ (`runState` 0) $ (state $ \s -> (10, 10)) `bind'` (\x -> state $ \s -> (x + 20, s)) -- (30,10)
  print $ (`runState` 5) $ return' 10 `bind'` (\_ -> get') `bind'` (\x -> return' x) -- (5,5)

File: monad/state.py
# -*- coding: utf-8 -*-


class State():
    @staticmethod
    def ret(x):
        return State(lambda s: (x, s))

    def __init__(self, f):
        self.f = f

    def __str__(self):
        return "State " + str(self.f)

    def bind(self, state):
        return State(lambda s:
                     (lambda r1, s1:
                      (lambda r2, s2:
                       (r2, s2))(*state(r1).runState(s1)))(*self.runState(s)))

    def get(self):
        return State(lambda s:
                     (lambda s1: (s1, s1))(self.execState(s)))

    def put(self, x):
        return State(lambda s: ((), x))

    def modify(self, f):
        return State(lambda s: ((), f(self.execState(s))))

    def runState(self, s):
        return self.f(s)

    def evalState(self, s):
        return self.runState(s)[0]

    def execState(self, s):
        return self.runState(s)[1]


if __name__ == '__main__':
    print(State
          .ret(10)
          .runState(0))  # (10, 0)

    print(State
          .ret(10)
          .bind(lambda x: State(lambda s: (-1, -1)))
          .runState(10))  # (-1, -1)

    print(State
          .ret(10)
          .bind(lambda x: State(lambda s: (-1, -1)))
          .bind(lambda x: State.ret(100))
          .runState(10))  # (100, -1)

    print(State
          .ret(10)
          .bind(lambda x: State(lambda s: (-1, s)))
          .runState(10))  # (-1, 10)

    print(State
          .ret(1)
          .get()
          .bind(lambda x: State(lambda s: (-1, -1)))
          .modify(lambda x: x + 9)
          .bind(lambda x: State(lambda s: (100, s)))
          .runState(10))  # (100, 8)
